src/linux_android.rs

```diff
@@ -1,19 +1,68 @@
 //! Implementation for Linux / Android without `/dev/urandom` fallback
-use crate::{util_libc, Error};
+use crate::{util_unix::sys_fill_exact, Error};
 use core::mem::MaybeUninit;
 
 pub fn getrandom_inner(dest: &mut [MaybeUninit<u8>]) -> Result<(), Error> {
-    util_libc::sys_fill_exact(dest, getrandom_syscall)
+    sys_fill_exact(dest, getrandom_syscall)
 }
 
+// The value of `EINTR` is not architecture-specific. It is checked against
+// `libc::EINTR` by linux_android_with_fallback.rs.
+pub const EINTR: i32 = 4;
+
 // Also used by linux_android_with_fallback to check if the syscall is available.
-pub fn getrandom_syscall(buf: &mut [MaybeUninit<u8>]) -> libc::ssize_t {
-    unsafe {
-        libc::syscall(
-            libc::SYS_getrandom,
-            buf.as_mut_ptr().cast::<core::ffi::c_void>(),
-            buf.len(),
-            0,
-        ) as libc::ssize_t
+cfg_if! {
+    // TODO: Expand inilne assembly to other architectures to avoid depending
+    // on libc on Linux.
+    if #[cfg(target_arch = "x86_64")] {
+        type Word = u64;
+        type IWord = i64;
+
+        #[allow(non_upper_case_globals)]
+        pub const SYS_getrandom: IWord = if cfg!(target_arch = "x86_64") { 318 } else { unimplemented!() };
+
+        pub fn getrandom_syscall(buf: &mut [MaybeUninit<u8>]) -> Result<usize, Error> {
+            const _:() = assert!(core::mem::size_of::<Word>() == core::mem::size_of::<usize>());
+
+            let mut ret: IWord;
+            let flags = 0;
+            unsafe {
+                core::arch::asm!(
+                    "syscall",
+                    inout("rax") SYS_getrandom => ret,
+                    in("rdi") buf.as_mut_ptr(),
+                    in("rsi") buf.len(),
+                    in("rdx") flags,
+                    lateout("rcx") _,
+                    lateout("r11") _,
+                    options(nostack),
+                );
+            }
+            match Word::try_from(ret) {
+                Ok(written) => {
+                    const _:() = assert!(core::mem::size_of::<Word>() <= core::mem::size_of::<usize>());
```

Comment:
Depending on whether/how #463 is resolved, we may need to `unpoison(dest, written)` here.